added NNX vs JAX transformations guide (#3819)
google/flax: docs/experimental/nnx/transforms.rst

NNX vs JAX Transformations
==========================

In this guide, you will learn the differences using NNX and JAX transformations, and how to
seamlessly switch between them or use them together. We will be focusing on the ``jit`` and
``grad`` function transformations in this guide.

First, let's set up imports and generate some dummy data:

.. testcode::

  from flax.experimental import nnx
  import jax

  x = jax.random.normal(jax.random.key(0), (1, 2))
  y = jax.random.normal(jax.random.key(1), (1, 3))

Differences between NNX and JAX transformations
***********************************************

The primary difference between NNX and JAX transformations is that NNX transformations allow you to

> transform functions that take in NNX objects as arguments, even those with mutable internal state,

Reviewer: ```suggestion
transform functions that take in NNX graph objects as arguments (`Module`, `Rngs`, `Optimizer`, etc), even those whose state will be mutated,
```

whereas they aren't recognized in JAX transformations. Therefore NNX transformations can transform
functions that are not pure and make mutations and side-effects.

> The NNX API allows the user to split a stateful NNX object into a ``State`` object that contains the
> internal variables and a ``GraphDef`` object that contains the NNX tree structure. These two NNX
> classes are registered as pytrees in JAX and so JAX will recognize them when applying JAX
> transformations to functions that accept those objects as arguments.
> 
> The ``State`` and ``GraphDef`` objects can be merged to generate the stateful NNX object once again
> to be passed into NNX transformations. Using the NNX API of splitting and merging, the user can
> seamlessly switch between stateful and stateless objects and use both NNX and JAX transformations.

Reviewer: ```suggestion
NNX's `Functional API <https://flax.readthedocs.io/en/latest/experimental/nnx/nnx_basics.html#the-functional-api>`_ provides a way to convert graph structures to pytrees and back, by doing this at every function boundary you can effectively use graph structures with any JAX transform and propagate state updates in a way consistent with functional purity. NNX custom transforms such as ``nnx.jit`` and ``nnx.grad`` simply remove the boilerplate, as a result the code looks stateful.
```

``State`` and ``GraphDef`` objects and must return an updated copy of them to maintain the purity of
the transformed function.

.. codediff::
  :title_left: NNX transforms
  :title_right: JAX transforms
  :sync:

  @nnx.jit
  def train_step(model, x, y):
    def loss_fn(model):
      return ((model(x) - y) ** 2).mean()
    grads = nnx.grad(loss_fn)(model)

    model.update(
      jax.tree_util.tree_map(
        lambda p, g: p - 0.1 * g, model.extract(nnx.Param), grads
      )
    )

  model = nnx.Linear(2, 3, rngs=nnx.Rngs(0))
  train_step(model, x, y)

  ---
  @jax.jit #!
  def train_step(static, state, x, y): #!
    def loss_fn(static, state): #!
      model = static.merge(state) #!
      return ((model(x) - y) ** 2).mean()
    grads = jax.grad(loss_fn, argnums=1)(static, state) #!

    model = static.merge(state) #!
    model.update(
      jax.tree_util.tree_map(
        lambda p, g: p - 0.1 * g, model.extract(nnx.Param), grads
      )
    )
    return model.split() #!

  static, state = nnx.Linear(2, 3, rngs=nnx.Rngs(0)).split() #!
  static, state = train_step(static, state, x, y) #!

> 

Reviewer: ```suggestion
mode.update(state) #!
```

Mixing NNX and JAX transformations
**********************************

NNX and JAX transformations can be mixed together, so long as the JAX-transformed function is
pure and has valid argument types that are recognized by JAX.

.. codediff::
  :title_left: Using ``nnx.jit`` with ``jax.grad``
  :title_right: Using ``jax.jit`` with ``nnx.grad``
  :sync:

  @nnx.jit
  def train_step(model, x, y):
    def loss_fn(static, state): #!
      model = static.merge(state)
      return ((model(x) - y) ** 2).mean()
    grads = jax.grad(loss_fn, 1)(*model.split()) #!

    model.update(
      jax.tree_util.tree_map(
        lambda p, g: p - 0.1 * g, model.extract(nnx.Param), grads
      )
    )

  model = nnx.Linear(2, 3, rngs=nnx.Rngs(0))
  train_step(model, x, y)

  ---
  @jax.jit #!

>   def train_step(static, state, x, y): #!
>     def loss_fn(model):
>       return ((model(x) - y) ** 2).mean()
>     model = static.merge(state)
>     grads = nnx.grad(loss_fn)(model)

Reviewer: suggest merging at the beginning of the function
```suggestion
  def train_step(static, state, x, y): #!
    model = static.merge(state)
    def loss_fn(model):
      return ((model(x) - y) ** 2).mean()
    grads = nnx.grad(loss_fn)(model)
```
